src/System.Collections.Immutable/src/System/Collections/Immutable/ImmutableExtensions.cs

```diff
@@ -85,6 +85,38 @@ internal static int GetCount<T>(ref IEnumerable<T> sequence)
 
             return count;
         }
+        
+        /// <summary>
+        /// Copies part of an IEnumerable&lt;T&gt; to an array, starting at the specified index.
+        /// </summary>
+        /// <typeparam name="T">The type of element.</typeparam>
+        /// <param name="sequence">The sequence of elements you want to copy.</param>
+        /// <param name="dest">The destination array.</param>
+        /// <param name="index">The index in the array to start copying to.</param>
+        internal static void CopyTo(this IEnumerable<T> sequence, T[] dest, int index)
+        {
+            var collectionOfT = items as ICollection<T>;
+            
+            if (collectionOfT != null)
+            {
+                collectionOfT.CopyTo(tmp, index);
```

Comment:
Not unreliabile, just with a different ordering between what CopyTo does and what enumeration does.  Consider a collection like (pseudo-code):
```C#
public class MyStack<T> : IEnumerable<T>, ICollection
{
    private T[] _array;
    private int _count;
    ...
    public void CopyTo(T[] array, int index)
    {
        Array.Copy(_array, 0, array, index, _array.Length);
    }

    public IEnumerable<T> GetEnumerable()
    {
        for (int i = _count - 1; i >= 0; i--)
        {
            yield return _array[i];
        }
    }
}
```
It's implemented so that the order of the elements that end up in the array is different if you use CopyTo versus enumerating it to fill in the array.  I don't see anything in the documentation for ICollection that dictates the order in which the elements are copied, that it must match the ordering provided by the IEnumerable implementation.

Now, maybe in practice this doesn't happen, or maybe even if it does it's so rare that it's acceptable risk.  I was pointing out that it can be an observable behavior change.